### PR TITLE
Add some automated testing

### PR DESCRIPTION
**Pull request by nhaehnle**
The tests are integrated into the build system via `check-llvm-dialects`.

A next step would be to add a GitHub CI action. We could re-use the amdvlk docker images perhaps, or use snapshots from https://apt.llvm.org/?

### CODE REVIEW

**Review by jasilvanus — COMMENTED** [review]
Nice, thanks!

Some inline comments, most of which actually apply to the generated source and not the added testing, sorry for the distraction. Feel free address those later.

File: test/example/generated/ExampleDialect.h.inc [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+} // namespace llvm_dialects
+
+namespace xd {
+
+
+    class ExampleDialect : public ::llvm_dialects::DialectImpl<ExampleDialect> {
+      friend DialectImpl;
+      friend ::llvm_dialects::DialectContext;
+
+      void anchor() override;
```
[the remark this answers is not in the document]
Comment (nhaehnle) [reply test/example/generated/ExampleDialect.h.inc:24]: It ensures the *placement* of the vtable for linking. At least in the Itanium ABI, the compiler emits the vtable when it emits the definition of the first explicit virtual method (if one exists).

Having the anchor, and putting its definition in a .cpp file, is a simple pattern that guarantees that the vtable is only emitted once. I think this mostly saves compile and link time, maybe it has even more impact on Windows when DLL linking comes into play.

This is a pattern that's used a bunch in LLVM.

File: test/example/generated/ExampleDialect.h.inc [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+      static Key& getKey();
+
+    private:
+      ExampleDialect(::llvm::LLVMContext& context);
+
+      static ::llvm_dialects::Dialect* make(::llvm::LLVMContext& context);
+  
+      public:
+        ::llvm::AttributeList getAttributeList(size_t index) const {
```
[the remark this answers is not in the document]
Comment (nhaehnle) [reply test/example/generated/ExampleDialect.h.inc:34]: It ought to only be used from generated code, but sure why not. Separate change, though :)

File: test/lit.cfg.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+config.test_format = lit.formats.ShTest(not llvm_config.use_lit_shell)
+
+# suffixes: A list of file extensions to treat as test files. This is overridden
+# by individual lit.local.cfg files in the test subdirectories.
+config.suffixes = ['.test']
+
+# excludes: A list of directories to exclude from the testsuite. The 'Inputs'
+# subdirectories contain auxiliary inputs for various tests in their parent
+# directories.
+config.excludes = ['Inputs', 'CMakeLists.txt', 'README.txt', 'LICENSE.txt']
```
[the remark this answers is not in the document]
Comment (nhaehnle) [reply test/lit.cfg.py:30]: No. I'm leaving the "Inputs" though just as a reminder that this feature exists.

File: test/example/generated/ExampleDialect.cpp.inc [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+{
+  ::llvm::AttrBuilder attrBuilder{context};
+attrBuilder.addAttribute(::llvm::Attribute::NoUnwind);
+attrBuilder.addMemoryAttr(::llvm::MemoryEffects(::llvm::MemoryEffects::Location::InaccessibleMem, ::llvm::ModRefInfo::ModRef));
+m_attributeLists[2] = ::llvm::AttributeList::get(context, ::llvm::AttributeList::FunctionIndex, attrBuilder);
+}
+}
+
+
+      const ::llvm::StringLiteral Add32Op::s_name{"xd.add32"};
```
[the remark this answers is not in the document]
Comment (nhaehnle) [reply test/example/generated/ExampleDialect.cpp.inc:52]: Good question. I think I originally had it as `constexpr` but in the header. Not sure why I moved it and made it `const`.

**Review by jasilvanus — APPROVED** [review]

**Pull request merged**